### PR TITLE
Add direct perturbation-based sensitivity coefficient calculation support in STM

### LINKED ISSUE #28671: Feature Request: Direct perturbation-based sensitivity analysis
### Motivation

Myself, and others in C110 are performing more and more uncertainty quantification (UQ) and sensitivity analysis (SA) in our work. From microreactors to space power reactors, even moving to GenIV designs, UQ/SA is becoming significantly more popular (thank goodness). We use the stochastic tools module (STM) frequently, and having additional methods of SA (we tend to use variance-based sensitivity analysis) would be a welcome addition. 

### Design

Perturbation-based SA provides a different method of analysis for expectation values of parameters of interest. More ways to report sensitivity would be great. In particular, the Nuclear Regulatory Commission (NRC) is beginning to take a big interest in SA, and we must be able to provide this for them. I have a charge number for FY25 which entails SA work on a GenIV reactor for the NRC.

### Impact

This will add something new to STM. 

@grmnptr

### PR DESCRIPTION
**Pull request by grmnptr**
Closes #28671

### CODE REVIEW

File: modules/stochastic_tools/doc/content/source/reporters/DirectPerturbationReporter.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+\begin{equation}
+  f_1(\mu_1,\mu_2,\mu_2)=\mu_1+2\mu_2+3\mu_3
+\end{equation}
+
+\begin{equation}
+  f_2(\mu_1,\mu_2,\mu_2)=4\mu_1+5\mu_2+6\mu_3
+\end{equation}
+
+\begin{equation}
+  f_2(\mu_1,\mu_2,\mu_2)=7\mu_1+8\mu_2+9\mu_3
```
Comment (GiudGiud) [inline modules/stochastic_tools/doc/content/source/reporters/DirectPerturbationReporter.md:26]:
```suggestion
  f_3(\mu_1,\mu_2,\mu_2)=7\mu_1+8\mu_2+9\mu_3
```

File: modules/stochastic_tools/doc/content/source/samplers/DirectPerturbationSampler.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+parameter.
+The implementation is fully parallelized meaning that the model
+evaluations and the sensitivity calculation all happen in
+parallel.
+
+
+## Example Input Syntax
+
+the following input generates perturbed samples for a
+3-dimesnionsal parameter space around a nominal value
```
Comment (GiudGiud) [inline modules/stochastic_tools/doc/content/source/samplers/DirectPerturbationSampler.md:68]:
```suggestion
3-dimensional parameter space around a nominal value
```

File: modules/stochastic_tools/doc/content/source/samplers/DirectPerturbationSampler.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+!syntax description /Samplers/DirectPerturbationSampler
+
+## Overview
+
+The direct perturbation method is the simplest approach for
+computing local sensitivity coefficients using the finite
+difference approximation of the derivatives at a certain
+point in the parameter domain. The coordinates of the reference
+location can be supplied through the [!param](/Samplers/DirectPerturbationSampler/nominal_parameter_values) parameter.
+The parameter vector holding the paraemters can be denoted by:
```
Comment (GiudGiud) [inline modules/stochastic_tools/doc/content/source/samplers/DirectPerturbationSampler.md:12]:
```suggestion
The parameter vector holding the parameters can be denoted by:
```

File: modules/stochastic_tools/doc/content/source/samplers/DirectPerturbationSampler.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+$\vec{\mu}=\left[\mu_1,\mu_2,\mu_3,...,\mu_{N_p}\right]^T$, where
+$N_p$ is the number of paraemters. The direct perturbation-based
```
Comment (GiudGiud) [inline modules/stochastic_tools/doc/content/source/samplers/DirectPerturbationSampler.md:14]:
```suggestion
$N_p$ is the number of parameters. The direct perturbation-based
```

File: modules/stochastic_tools/include/reporters/DirectPerturbationReporter.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+#include "GeneralReporter.h"
+#include "VectorCalculators.h"
+
+#include "nlohmann/json.h"
+
+// Forward declaration
+class DirectPerturbationSampler;
+
+class DirectPerturbationReporter : public GeneralReporter
```
Comment (GiudGiud) [inline modules/stochastic_tools/include/reporters/DirectPerturbationReporter.h:20]: Docstring

File: modules/stochastic_tools/include/reporters/DirectPerturbationReporter.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  /// Direct perturbation sampler
+  DirectPerturbationSampler & _sampler;
+
+  /// Whether or not initialize() has been called for reporter value declaration
+  bool _initialized;
+};
+
+/**
+ * Reporter context for computing direct perturbation-based sensitivity
```
Comment (GiudGiud) [inline modules/stochastic_tools/include/reporters/DirectPerturbationReporter.h:46]: Docstring

File: modules/stochastic_tools/include/reporters/DirectPerturbationReporter.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+   * Constructor
+   * @param other A parallel object, usually the MooseApp
+   * @param producer The producer object for the reporter
+   * @param state A reporter state (a vector of some types in this case)
+   * @param sampler The sampler holding information on the direct perturbation paraemters
+   * @param data The data coming back from the executed models
+   */
+  DirectPerturbationReporterContext(const libMesh::ParallelObject & other,
+                                    const MooseObject & producer,
+                                    ReporterState<std::vector<DataType>> & state,
```
Comment (GiudGiud) [inline modules/stochastic_tools/include/reporters/DirectPerturbationReporter.h:63]: I think docstring is for another function

File: modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+registerMooseObject("StochasticToolsApp", DirectPerturbationReporter);
+
+InputParameters
+DirectPerturbationReporter::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addClassDescription("Compute local sensitivities using the direct perturbation method.");
+
+  params.addRequiredParam<SamplerName>("sampler",
+                                       "Direct PErturbation sampler used to generate samples.");
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:22]:
```suggestion
                                       "Direct Perturbation sampler used to generate samples.");
```

File: modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+  params.addParam<std::vector<VectorPostprocessorName>>(
+      "vectorpostprocessors",
+      "List of VectorPostprocessor(s) to utilized for sensitivity computations.");
+  params.addParam<std::vector<ReporterName>>(
+      "reporters", {}, "List of Reporter values to utilized for sensitivity computations.");
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:27]:
```suggestion
      "reporters", {}, "List of Reporter values to utilize for statistic computations.");
```

File: modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+InputParameters
+DirectPerturbationReporter::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addClassDescription("Compute local sensitivities using the direct perturbation method.");
+
+  params.addRequiredParam<SamplerName>("sampler",
+                                       "Direct PErturbation sampler used to generate samples.");
+  params.addParam<std::vector<VectorPostprocessorName>>(
+      "vectorpostprocessors",
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:24]: Add empty {} there too for simplicity

File: modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+  return params;
+}
+
+DirectPerturbationReporter::DirectPerturbationReporter(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    _sampler(getSampler<DirectPerturbationSampler>("sampler")),
+    _initialized(false)
+{
+  if ((!isParamValid("reporters") && !isParamValid("vectorpostprocessors")) ||
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:37]: Remove the paramvalid part now as the default {} are always valid

File: modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    if (!unsupported_types.empty())
+      paramError("reporters",
+                 "The following reporter value(s) do not have a type supported by the "
+                 "DirectPerturbationReporter:\n",
+                 MooseUtils::join(unsupported_types, ", "));
+  }
+
+  // Sensitivities from VPP
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:72]: Remove once {} default

File: modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+      if constexpr (std::is_arithmetic<InnerValueType>::value &&
+                    !std::is_same<InnerValueType, bool>::value)
+      {
+        // Iterate over each inner vector in the outer vector
+        for (auto & inner_index : index_range(add_to))
+        {
+          mooseAssert(add_to[inner_index].size() == to_add[inner_index].size(),
+                      "The vectors for summation have different sizes!");
+          for (const auto index : index_range(add_to[inner_index]))
+            add_to[inner_index][index] += to_add[inner_index][index] / interval;
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:201]: Need more of these in the nested if conditions

File: modules/stochastic_tools/test/tests/reporters/directperturbation/tests [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  [central_difference]
+    requirement = 'The system shall support the ability to compute direct perturbation-based sensitivity values with central differencing '
+    [distributed]
+      type = JSONDiff
+      input = dp_main.i
+      jsondiff = dp_main_out.json
+      allow_test_objects = true
+      detail = 'on distributed data sets;'
+      min_parallel = 4
+      max_parallel = 4
```
Comment (GiudGiud) [inline modules/stochastic_tools/test/tests/reporters/directperturbation/tests:14]: Why those restrictions?

File: modules/stochastic_tools/test/tests/samplers/directperturbation/tests [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+      csvdiff = 'directperturbation_out_data_0001.csv'
+      detail = "with replicated output;"
+    []
+    [parallel]
+      type = CSVDiff
+      input = directperturbation.i
+      cli_args = 'VectorPostprocessors/data/parallel_type=DISTRIBUTED Outputs/file_base=dp_cd_3'
+      min_parallel = 3
+      max_parallel = 3
+      csvdiff = 'dp_cd_3_data_0001.csv.0 dp_cd_3_data_0001.csv.1 dp_cd_3_data_0001.csv.2'
```
Comment (GiudGiud) [inline modules/stochastic_tools/test/tests/samplers/directperturbation/tests:20]: Like here I understand you are stuck with 3 processes. But for reporters?

File: modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C (hunk 13 above)
Comment (GiudGiud) [reply modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:201]: by these I meant
```
    static_assert(Moose::always_false<DataType>,
                  "Sensitivity coefficient computation is not implemented for the given type!");
```

Comment (grmnptr) [reply modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C:201]: Yeah addded, will push soon! Very nice catch!

File: modules/stochastic_tools/test/tests/reporters/directperturbation/tests (hunk 14 above)
Comment (grmnptr) [reply modules/stochastic_tools/test/tests/reporters/directperturbation/tests:14]: Aaaa, right! I forgot that we already have a test with `-p 2`. That should actually be enough for this. Removing this.

File: modules/stochastic_tools/test/tests/samplers/directperturbation/tests (hunk 15 above)
Comment (grmnptr) [reply modules/stochastic_tools/test/tests/samplers/directperturbation/tests:20]: Yeah removed that. That reporter will always output on root, and only the data used is distributed/root. Great catch!

**Review by GiudGiud — APPROVED** [review]

**Comment by grmnptr** [comment]
Added extra tests like valgrind an everything passes, openmpi failure is not related.

**Pull request merged**